tsa-core/src/main/kotlin/org/usvm/machine/interpreter/TvmDictOperationInterpreter.kt

```diff
@@ -1515,27 +1677,100 @@ class TvmDictOperationInterpreter(
+        if (dicts.size == 1) {
+            // avoid forking in this case
+            doDictGetImpl(
+                scope,
+                this@with,
+                keyLength,
+                dicts.single().second,
+                keyKind,
+                key,
+                nullDefaultValue,
+                inst,
+                valueType,
+            )
+        } else {
+            scope.doWithConditions(actions) { dictCellRef ->
+                doDictGetImpl(this, this@with, keyLength, dictCellRef, keyKind, key, nullDefaultValue, inst, valueType)
+            }
+        }
+    }
 
+    private fun doDictGetImpl(
+        scope: TvmStepScopeManager,
+        ctx: TvmContext,
+        keyLength: Int,
+        dictCellRef: UConcreteHeapRef,
+        keyKind: DictKeyKind,
+        key: UExpr<UBvSort>,
+        nullDefaultValue: Boolean,
+        inst: TvmInst,
+        valueType: DictValueType,
+    ) {
         val dictId = DictId(keyLength)
 
         assertDictKeyLength(scope, dictCellRef, keyLength)
             ?: return
-        assertDictIsNotEmpty(scope, dictCellRef, dictId)
-            ?: return
 
+        val keySort = scope.ctx.mkBvSort(keyLength.toUInt())
+        val allSetEntries =
+            scope.calcOnStateCtx {
+                memory.setEntries(dictCellRef, dictId, keySort, DictKeyInfo)
+            }
+        val isInput =
+            allSetEntries.isInput ||
+                scope.calcOnState {
+                    this.inputDictionaryStorage.memory.keys
+                        .contains(dictCellRef)
+                }
+        if (!isInput) {
+            assertDictIsNotEmpty(scope, dictCellRef, dictId)
+                ?: return
+        }
         val dictContainsKey =
-            scope.calcOnState {
-                dictContainsKey(dictCellRef, dictId, key)
+            if (isInput) {
+                val dictConcreteRef = dictCellRef
+                val (inputDict, rootInputDictInfo) =
+                    scope.calcOnState {
+                        readInputDictionary(dictConcreteRef, keySort, keyKind)
+                    }
+                val keyExists = scope.calcOnState { makeSymbolicPrimitive(ctx.boolSort) }
+                val (cs, updatedRootInfo) =
+                    inputDict.doDictHasKey(
+                        scope.ctx,
+                        KeyType(key, keyKind),
+                        rootInputDictInfo,
+                        scope.calcOnState { makeFreshKeyConstant(keySort, keyKind) },
+                        keyExists,
+                    )
+                scope.assert(scope.ctx.mkAnd(cs))
+                    ?: return
+                scope.calcOnStateCtx {
+                    inputDictionaryStorage = inputDictionaryStorage.set(dictConcreteRef, inputDict, updatedRootInfo)
+                }
+                keyExists
+            } else {
+                scope.calcOnState {
+                    dictContainsKey(dictCellRef, dictId, key)
+                }
```

Comment:
It is really not obvious that this function can be called only on allocated dict. At least rename it, better to make a general function